src/main/java/de/jordanmruczynski/backend/service/CinemaService.java

```diff
@@ -0,0 +1,159 @@
+import jakarta.persistence.EntityNotFoundException;
+import jakarta.transaction.Transactional;
+import org.springframework.stereotype.Service;
+
+import java.math.BigDecimal;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+public class CinemaService implements ICinemaService {
+
+    private final ScreeningRepository screeningRepository;
+    private final ReservationRepository reservationRepository;
+    private final SeatRepository seatRepository;
+    private final TicketRepository ticketRepository;
+
+    public CinemaService(ScreeningRepository screeningRepository, ReservationRepository reservationRepository, SeatRepository seatRepository, TicketRepository ticketRepository) {
+        this.screeningRepository = screeningRepository;
+        this.reservationRepository = reservationRepository;
+        this.seatRepository = seatRepository;
+        this.ticketRepository = ticketRepository;
+    }
+
+
+    @Override
+    public List<MovieScreeningsDTO> getAllScreenings(LocalDateTime start, LocalDateTime end) {
+        List<Screening> screenings = screeningRepository.findAllByScreeningStartTimeBetween(start, end);
+
+        screenings.sort(
+                Comparator.comparing((Screening s) -> s.getMovie().getTitle())
+                        .thenComparing(Screening::getScreeningStartTime)
+        );
+
+        final Map<String, List<LocalDateTime>> screeningsMap = screenings.stream()
+                .collect(Collectors.groupingBy(
+                        screening -> screening.getMovie().getTitle(),
+                        Collectors.mapping(Screening::getScreeningStartTime, Collectors.toList())
+                ));
+
+        return screeningsMap.entrySet().stream()
+                .sorted(Map.Entry.comparingByKey())
+                .map(entry -> new MovieScreeningsDTO(entry.getKey(), entry.getValue()))
+                .collect(Collectors.toList());
+    }
+
+
+    @Override
+    @Transactional
+    public ReservationResponse createReservation(ReservationRequest request) {
+        Screening screening = screeningRepository.findById(request.screeningId())
+                .orElseThrow(() -> new EntityNotFoundException("Screening not found"));
+
+        LocalDateTime now = LocalDateTime.now();
+        if (Duration.between(now, screening.getScreeningStartTime()).toMinutes() < 15) {
+            throw new ReservationTooLateException("Reservations are only allowed up to 15 minutes before screening start time.");
+        }
+
+
+        List<Seat> selectedSeats = request.tickets().stream()
+                .map(ticketRequest -> seatRepository.findById(ticketRequest.seatId())
+                        .orElseThrow(() -> new ResourceNotFoundException("Seat not found")))
+                .collect(Collectors.toList());
+
+        Set<Seat> reservedSeats = ticketRepository.findReservedSeatsByScreening(screening);
```

Comment:
Wyciągamy wszystkie zajęte miejsca, żeby później sprawdzić, czy są wśród nich miejsca z rezerwacji. Myślę, że można to uprościć...

Reply:
```java
Set<Integer> selectedSeatIds = request.tickets().stream()
        .map(TicketRequest::seatId)
        .collect(Collectors.toSet());

Set<Integer> reservedSeatIds = ticketRepository.findReservedSeatIdsByScreeningAndSeatIds(screening.getId(), selectedSeatIds);
```
Obstawiam, że dałoby się to jeszcze jakoś bardziej uprościć od strony bazy danych, ale nie przychodzi mi w tym momencie nic do głowy.